llama_hub/llama_packs/panel_chatbot/app.py

```diff
@@ -0,0 +1,380 @@
+        return index
+
+    @param.depends("status")
+    def _is_loading(self):
+        return self.status not in [INDEX_LOADED, INDEX_NOT_LOADED]
+
+    @param.depends("status")
+    def _is_not_loading(self):
+        return self.status in [INDEX_LOADED, INDEX_NOT_LOADED]
+
+    @param.depends("_load", watch=True)
+    async def load(self):
+        """Loads the repository index either from the cache or by downloading from
+        the repository"""
+        self._update_status("Loading ...")
+        self.value = None
+
+        docs = await self._get_docs()
+        self.value = await self._get_index(docs)
+        self._update_status(INDEX_LOADED)
+
+    @param.depends("_reload", watch=True)
+    async def reload(self):
+        self._update_status("Deleteing cached index ...")
+        if self._cached_docs_path.exists():
+            self._cached_docs_path.unlink()
+        if self._cached_index_path.exists():
+            self._cached_index_path.unlink()
+
+        await self.load()
+
+    def _update_status(self, text):
+        with param.edit_constant(self):
+            self.status = text
+        print(text)
+
+    @param.depends("owner", "repo")
+    def github_url(self):
+        """Returns a html string with a link to the github repository"""
+        text = f"{self.owner}/{self.repo}"
+        href = f"https://github.com/{text}"
+        return f"<a href='{href}' target='_blank'>{text}</a>"
+
+    @property
+    def index_exists(self):
+        """Returns True if the index already exists"""
+        return self._cached_docs_path.exists() and self._cached_index_path.exists()
+
+
+def powered_by():
+    """Returns a component describing the frameworks powering the chat ui"""
+    params = {"height": 40, "sizing_mode": "fixed", "margin": (0, 10)}
+    return pn.Column(
+        pn.pane.Markdown("### AI Powered By", margin=(10, 5, 10, 0)),
+        pn.Row(
+            pn.pane.Image(LLAMA_INDEX_LOGO, link_url=LLAMA_INDEX_URL, **params),
+            pn.pane.Image(CHAT_GPT_LOGO, link_url=CHAT_GPT_URL, **params),
+            pn.pane.Image(PANEL_LOGO[pn.config.theme], link_url=PANEL_URL, **params),
+            align="center",
+        ),
+    )
+
+
+async def chat_component(index: VectorStoreIndex, index_loader: IndexLoader):
+    """Returns the chat component powering the main area of the application"""
+    if not index:
+        return pn.Column(
+            pn.chat.ChatMessage(
+                "You are a now a *GitHub Repository assistant*.",
+                user="System",
+            ),
+            pn.chat.ChatMessage(
+                "Please **load a GitHub Repository** to start chatting with me. This can take from seconds to minutes!",
+                user="Assistant",
+            ),
+        )
+
+    chat_engine = index.as_chat_engine(chat_mode="context", verbose=True)
+
+    async def generate_response(contents, user, instance):
+        response = chat_engine.stream_chat(contents)
```

Comment:
I wanted to use `astream_chat` here. I tried looking at the documentation. I tried different ways of using it. But I could not figure out how to get it working.

Reply:
oh, do you mind sharing here quickly what issues/tracebacks you were running into?

Reply:
None. It was just not working. Probably because i dont know how to write the code. Do you have a working example you could share?

Reply:
here's one example: https://docs.llamaindex.ai/en/stable/examples/agent/openai_agent_parallel_function_calling.html

maybe something we can consider for a future update on this.

Reply:
This async examples seems to be working

```python


import nest_asyncio

from llama_index import SimpleDirectoryReader
from llama_index import VectorStoreIndex
import panel as pn

nest_asyncio.apply()

@pn.cache
def get_index()->VectorStoreIndex:
    print("Creating index ...")
    reader = SimpleDirectoryReader(
        input_files=[__file__]
    )
    docs = reader.load_data()
    return VectorStoreIndex.from_documents(docs, use_async=True)

index: VectorStoreIndex = get_index()

chat_engine = index.as_chat_engine(chat_mode="context", verbose=True)

async def generate_response(contents, user, instance):
    response = await chat_engine.astream_chat(contents)
    text = ""
    async for token in response.async_response_gen():
        text += token
        yield text

pn.chat.ChatInterface(
    callback=generate_response,
    sizing_mode="stretch_both",
).servable()
```

But I experience issues in larger, "real" applications will try to track down the issue and report in a separate issue as its not affecting this PR.

Reply:
I've now change the response callback to use `.astream_chat` as above.